---
title: "How to extract coordinates and sequence from a fasta file"
output: html_document
date: "2024-06-04"
always_allow_html: true
editor_options: 
  chunk_output_type: console
---

## Installing Rust

First you need to have an updated Rust installation. Go to this [site](https://www.rust-lang.org/tools/install) to learn how to install Rust.


## Installing rustytools

You will need to have the devtools package installed...

```{r, eval=F}
devtools::install_github("scfurl/rustytools", force=T)
```


## Running a tool to extract consensus sequence data

Loading libraries
```{r, dpi=300, fig.height=4, fig.width = 6}
rm(list=ls())
suppressPackageStartupMessages({
  library(rustytools)
})

```

## Get consensus from fasta

This function will process a fasta file that has been processed using a tool such as samtools consensus.  The function will find all regions (subsequences) of the fasta that have sequence (A, C, T, G) and output a granges object containing the features and their sequences.  It is invoked as follows:

```{r}
root<-file.path(.libPaths(), "rustytools/extdata")
tfasta=file.path(root, "test.fa")
results<-get_consensus(tfasta, cores=1, split=1)
results[[1]]
```

In the above toy example we use the split argument to specify how many chunks to break the sequence into for parallel processing.  In this case it doesn't make sense to split the sequence into chunks because the example only has about 500-600 bases for each contig.  When running this on real fasta sequence from a large genome, one would want to leverage parallel processing.  In this next block we show how to run get_consensus on a big fasta.  One can specify the name of the genome using the genome argument.  For hg38, we find that this takes about 1-3 minutes per chromosome when using 16 cores on an intel iMac.  We recommend splitting the fasta contigs into 10000 chunks as below.

```{r, eval=F}
fasta="abigfasta.fa"
#fasta="~/Desktop/GENDX_HLA_Mix1_S1_L001_t1000.fa"
results<-get_consensus(fasta, cores=16, split=10000, genome="hg38")
```

Before running a full fasta on all contigs, you can check the first n contigs of a full sized fasta using the argument test_with_n. Note that the full fasta will be read in to ensure appropriate seqinfo data
```{r, eval=F}
fasta="abigfasta.fa"
#fasta="~/Desktop/GENDX_HLA_Mix1_S1_L001_t1000.fa"
results<-get_consensus(fasta, cores=16, split=10000, genome="hg38", test_with_n = 1)
```


## Appendix
```{r Appendix}
sessionInfo()
```
